conv_model_bwd.cpp

```diff
@@ -201,13 +219,15 @@ int conv_benchmark(int argc, char** argv) {
   if ((R == 1 && S == 1) ||
       (avoid_rim_fmas == 1) ||
       (non_1x1_with_strides == 1)) {
-    auto gemm_n = ofw;
+    auto w_gemm_pixels = ofw/w_block;
+    auto gemm_n = (w_gemm_pixels +  2 * pad_w) * (h_in_gemm - 2) + 2 * (w_gemm_pixels + pad_w);
+    //auto gemm_n = ofw;
     auto gemm_m = bc;
     auto gemm_k = bk;
     auto l_shape = libxsmm_create_gemm_shape( gemm_m, gemm_n, gemm_k, bc, bk, stride_w*bc, dtype, dtype, dtype, dtype );
     auto l_prefetch_flags = LIBXSMM_GEMM_PREFETCH_NONE;
     auto l_brconfig = libxsmm_create_gemm_batch_reduce_config( LIBXSMM_GEMM_BATCH_REDUCE_STRIDE, R*S*bc*bk*sizeof(DType), bk*ofhp*ofwp*sizeof(DType), Kb_step );
-    auto l_unary_shape = libxsmm_create_meltw_unary_shape(bc*ifwp, 1, bc*ifwp, bc*ifwp, dtype, dtype, dtype);
+    auto l_unary_shape = libxsmm_create_meltw_unary_shape(bc*ifwp/w_block, 1, bc*ifwp/w_block, bc*ifwp/w_block, dtype, dtype, dtype);
```

Comment:
I think this is where I get problems originally, not accounting for w_block when zeroing the rim. With this extra `/w_block factor it works just fine.

Reply:
To be precise this kernel is not used for rim zeroing but for the initial pixel zeroing. It seems to me still needs some massaging to account the case with h_in_gemm != 1 ?

Reply:
Right (x2).  Introduced `w_zero_pixels` and `zero_n` (similar to what we have but for ifw instead of ofw).

Reply:
Looks good to me, thanks! It would be great if your could check-in here some rudimentary standalone tests/shapes (that you are using in your standalone C++ driver?) to confirm that we are not missing corner cases in terms of correctness

Reply:
I don't have a good standalone test for parlooper (rather just test the problematic cases I see from PT extension code runs), but I can surely add one for resnet-50 conv sizes + a few tuning param combinations. Soon-ish.

Reply:
Failed cases:
1) ./conv_bwd "Abcdefg" 28 14 14 1024 2048 1 1 2 2 0 0 32 32 1 **2** 1 1 1 10 (both fp32 and bf16)
Do you think this can be expected? I'd think it is something with divisibility by 2 as sizes like
./run_conv_bwd.sh "Abcdefg" 28 14 14 1024 2048 1 1 2 2 0 0 32 32 1 **7** 1 1 1 10
work just fine.
2) ./conv_bwd "Abcdefg" 28 7 7 512 512 3 3 1 1 1 1 32 32 1 1 1 1 7 10 (both fp32 and bf16)
This one goes into `avoid_rim_fmas=1` so I guess it should not work with `h_in_gemm != 1`, right?

Reply:
> I don't have a good standalone test for parlooper (rather just test the problematic cases I see from PT extension code runs), but I can surely add one for resnet-50 conv sizes + a few tuning param combinations. Soon-ish.

Yep, no rush, it is not urgent

Reply:
> Failed cases:
> 
> 1. ./conv_bwd "Abcdefg" 28 14 14 1024 2048 1 1 2 2 0 0 32 32 1 **2** 1 1 1 10 (both fp32 and bf16)
>    Do you think this can be expected? I'd think it is something with divisibility by 2 as sizes like
>    ./run_conv_bwd.sh "Abcdefg" 28 14 14 1024 2048 1 1 2 2 0 0 32 32 1 **7** 1 1 1 10
>    work just fine.
> 2. ./conv_bwd "Abcdefg" 28 7 7 512 512 3 3 1 1 1 1 32 32 1 1 1 1 7 10 (both fp32 and bf16)
>    This one goes into `avoid_rim_fmas=1` so I guess it should not work with `h_in_gemm != 1`, right?

For 1: Yes, because ofw will be 7 and w_block should perfectly divide it, then the only "allowed" values for such cases would be 1 and 7 (the latter with horrible performance). Of course we could do fancier things in the code with multiple gemms with varying "n" such that we will not require "perfect" divisibility anymore, but this is future work :)

For 2: Also yes, avoid_rim_fmas algo per definition works *at most*  in one "row" at a time, as such h_in_gemm must be 1.

Reply:
Yes, thanks. Removed test cases for 1. and added an early exit for 2. with a verbose message. The test for conv_bwd is now ready, is this resembling something you had in mind when you suggested it?

Reply:
Looks good to me, thanks!